src/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.DivRem.cs

```diff
@@ -128,203 +129,194 @@ public unsafe static uint[] Remainder(uint[] left, uint[] right)
-                               remainder, remainderLength,
-                               shift);
-                }
+                // We have the digit!
+                if (bitsLength != 0)
+                    bits[n] = (uint)digit;
+                if (i < leftLength)
+                    left[i] = 0;
             }
         }
 
         [SecuritySafeCritical]
-        private unsafe static void LeftShift(uint* value, int valueLength,
-                                             uint* target, int targetLength,
-                                             int shift)
+        private unsafe static uint AddDivisor(uint* left, int leftLength,
+                                              uint* right, int rightLength)
         {
-            Debug.Assert(valueLength >= 1);
-            Debug.Assert(targetLength == valueLength ||
-                         targetLength == valueLength + 1);
-            Debug.Assert(shift >= 0 && shift < 32);
+            Debug.Assert(leftLength >= 0);
+            Debug.Assert(rightLength >= 0);
+            Debug.Assert(leftLength >= rightLength);
 
-            if (shift > 0)
-            {
-                int backShift = 32 - shift;
-                target[0] = value[0] << shift;
-                for (int i = 1; i < valueLength; i++)
-                {
-                    target[i] = (value[i] << shift)
-                        | (value[i - 1] >> backShift);
-                }
-                if (targetLength > valueLength)
-                {
-                    target[valueLength] =
-                        value[valueLength - 1] >> backShift;
-                }
-            }
-            else
+            // Repairs the dividend, if the last subtract was too much
+
+            int i = 0;
+            ulong carry = 0L;
+
+            for (; i < rightLength; i++)
             {
-                for (int i = 0; i < valueLength; i++)
-                    target[i] = value[i];
+                ulong digit = (left[i] + carry) + right[i];
+                left[i] = (uint)digit;
+                carry = digit >> 32;
             }
+
+            return (uint)carry;
         }
 
         [SecuritySafeCritical]
-        private unsafe static void RightShift(uint* value, int valueLength,
-                                              uint* target, int targetLength,
-                                              int shift)
+        private unsafe static uint SubtractDivisor(uint* left, int leftLength,
+                                                   uint* right, int rightLength,
+                                                   ulong q)
         {
-            Debug.Assert(valueLength >= 1);
-            Debug.Assert(targetLength == valueLength);
-            Debug.Assert(shift >= 0 && shift < 32);
+            Debug.Assert(leftLength >= 0);
+            Debug.Assert(rightLength >= 0);
+            Debug.Assert(leftLength >= rightLength);
+            Debug.Assert(q <= 0xFFFFFFFF);
```

Comment:
It's ok, and you can leave it if you like.  I just worry about typos, accidentally having one too few or one too many Fs, etc.